fix(app,api): Display thermocycler profile cycles

sfoster1 opened the pull request:
We identified a simple bug with the app: it displays wonky numbers for thermocycler profile cycles:
![image (1)](https://github.com/user-attachments/assets/8005c0b6-c55e-4b48-bf50-7d90f91fa63b)

This wouldn't repeat anything at all, what's going on?

Well, it turns out that when Protocol Designer added thermocycler support, they wanted something a bit more in depth than the Python protocol API's "list of steps, number of times to repeat the cycle" format. They wanted users to be able to add single steps and cycles, in arbitrary order.

To make the two styles work with the engine, we made the engine's `thermocycler/runProfile` command take a flat list of steps - any of the structure of repeated commands was removed. In the app's `CommandText` display, we then had to fix up the way we displayed profiles to remove references to a `repetitions` value that now didn't exist... and we just didn't, instead setting the `repetitions` element to just be the number of steps in the profile. 

Fixing this ended up being a bit involved.

## summary
- ad172608c64c8d00844a7748ae282a1c8284b745 Make a new interface for the hardware controller `execute_profile` function of the thermocycler, since it's important that thermocycler cycles execute inside an asyncio worker so that they won't be interrupted by e.g. pausing. This new interface takes the PD style of protocol. It relies under the hood on the same function that actually sends stuff to the thermocycler, so there shouldn't be any functional execution changes.
- 335e25ad61ec36927060993a21342c20ec325c52 Make a new `thermocycler/runExtendedProfile` command in the protocol engine, that takes as its parameters a structured list of either steps or cycles, adhering to PD's expectations since the PAPI's expectations are a strict subset of PD. This implements its command by using the new hardware controller.
- e842a2a47da30a62437c093516292b48a676369b Use the new structured data to implement a command text for the new command that can render the equivalent of what PD makes (and, again, what the PAPI does, which is a strict subset)
- e842a2a47da30a62437c093516292b48a676369b Also make the old command text not use a "repetitions" keyword that doesn't exist
- 0043ab7cf281b4dff4277c1350831671232559ea Switch the PAPI to emitting the new command in api 2.21

## new UI
These UI changes will be on  https://s3-us-west-2.amazonaws.com/opentrons-components/rqa-2771-thermocycler-extended-profiles/index.html?path=/docs/app-molecules-command-commandtext--docs for the desktop and general text and https://s3-us-west-2.amazonaws.com/opentrons-components/rqa-2771-thermocycler-extended-profiles/index.html?path=/docs/app-molecules-command-command--docs for the ODD colored-background elements whenever they upload. Here's a screenshot of what they were like when the PR was opened:

`CommandText` of `thermocycler/runProfile` :
![runProfileCommandText](https://github.com/user-attachments/assets/dcad4ce8-2f0e-418d-8618-066a084dc832)

Note the change to the first line. This needs wordsmithing.

`CommandText` of `thermocycler/runExtendedProfile`:
![runProfileExtended](https://github.com/user-attachments/assets/6b1a1977-9200-4d78-84dc-2961d76cfff9)

Note the two-level rendering. The common case will probably be that there's only cycles

## todo
- [x] wordsmith
- ~[ ] add to PD?~ no, but structured in a way that it will be easy eventually
- [x] Test. Oh boy

Closes RQA-2771

sfoster1 committed:
feat(api): add tc extended profiles to hcapi

These extended profiles match the way that you can configure a
thermocycler in protocol designer. They support a mix of cycles (more
than one per profile) and out-of-cycle steps, while keeping structure
down to hardware. That last part is sadly important to support
currently-exposed details up from the hardware controller.

sfoster1 committed:
feat(api,shared-data): add runExtendedProfile

This is a thermocycler command that accepts a structured sequence of
temperatures and wait times instead of a flat list.

A profile is a list of either steps or cycles; cycles are a repetition
count plus a list of steps.

The actual _functionality_ of the command is exactly the same as
thermocycler/runProfile; the only change is pushing down the structure
of the information, which is currently lost when creating a runProfile
command and thus unavailable to protocol clients for display.

sfoster1 committed:
fix(app): update thermocycler command text

This PR does a couple things with thermocycler command text.

First, it changes the phrasing of the thermocycler/runProfile command
text to make it more clear that no cycles are involved: it's just a big
list of steps.

Second, it adds a command text for the new
thermocycler/runExtendedProfile command, which does have the structured
data necessary to render things as cycles. This unfortunately makes it a
multilevel list, thus worsening the annoying problems with it.

Closes RQA-2771

sfoster1 committed:
fix(api): use extended profiles in PAPI

Use the new thermocycler/runExtendedProfile to implement
ThermocyclerContext.execute_profile() at or above protocol api version
2.21.

This new version of the command preserves the structure of the thermal
cycles and will allow the desktop app and ODD to present appropriate
information to the user.

sfoster1 committed:
wrong import

sfoster1 committed:
fix test and add new entry

sfoster1 committed:
fixup: wrong type in s-d commands

sfoster1 committed:
fixup: wrong semantics in api translation

sfoster1 commented:
<img width="926" alt="Screenshot 2024-10-04 at 9 47 38 AM" src="https://github.com/user-attachments/assets/fcd26290-5e54-4e44-82c7-d570a5a29f1c">

After those latest fixups, a protocol like this displays as above:
```python
requirements = {
    "robotType": "Flex",
    "apiLevel": "2.21"
}

def run(ctx):
    tr = ctx.load_labware('opentrons_96_tiprack_1000ul', 'D2')
    instr = ctx.load_instrument('flex_1channel_1000', 'left', tip_racks=[tr])
    tc = ctx.load_module('thermocyclerModuleV2')
    tc.load_labware("nest_96_wellplate_100ul_pcr_full_skirt")
    tc.close_lid()
    tc.execute_profile(
        steps=[{'temperature': 30, 'hold_time_seconds': 10},
         {'temperature': 50, 'hold_time_minutes': 1},
         {'temperature': 60, 'hold_time_seconds': 15}],
        repetitions=5,
        block_max_volume=100
    )
    instr.pick_up_tip()
    instr.return_tip()

```

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/thermocycler/run_extended_profile.py:
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python3
```
Accidental shebang?

sfoster1 replied:
yurp

SyntaxColoring commented on api/src/opentrons/hardware_control/modules/thermocycler.py:
```diff
@@ -363,6 +363,39 @@ async def cycle_temperatures(
+                self._total_step_count += 1
+                self._total_cycle_count += 1
+        task = self._loop.create_task(self._execute_profile(profile, volume))
+        self.make_cancellable(task)
```
For my edification, what's all this `make_cancellable()` stuff about? If we had just done `await self._execute_profile(profile_volume)`, our asyncio task could still be cancelled anyway, couldn't it?

sfoster1 replied:
That eventually calls [`ExecutionManager.register_cancellable_task`](https://github.com/Opentrons/opentrons/blob/66cfff9ac2173193939da288a01f9674dfb4a79b/api/src/opentrons/hardware_control/execution_manager.py#L69-L71) which means that when something calls [`ExecutionManager.cancel`](https://github.com/Opentrons/opentrons/blob/66cfff9ac2173193939da288a01f9674dfb4a79b/api/src/opentrons/hardware_control/execution_manager.py#L51-L58) as [protocol engine does](https://github.com/Opentrons/opentrons/blob/66cfff9ac2173193939da288a01f9674dfb4a79b/api/src/opentrons/protocol_engine/protocol_engine.py#L373) the task will get cancelled instead of orphaned. I think this might not be quite as necessary now that it's awaited, but a lot of these module asyncio tasks were implemented with an eye towards running asynchronously so they do all this stuff.

SyntaxColoring reviewed (approved):
The backend bits all look good to me, as discussed in Slack and on a call. TY!

sfoster1 commented:
<img width="821" alt="Screenshot 2024-10-04 at 10 48 26 AM" src="https://github.com/user-attachments/assets/ea229657-9649-4cc8-b258-3c923fb52386">
The same protocol but manually modified to use the new command. Here's the protocol; it implements the same thing as the previous protocol but using the real command. Don't try and view this in PD
[tc-test-extended.json](https://github.com/user-attachments/files/17260291/tc-test-extended.json)

sfoster1 committed:
fixup: remove shebang

mjhuff reviewed (approved):
App side looks good! Thanks for tackling this one...definitely not the innocuous UI bug as it first seemed.

Oh yeah, and thanks for adding the new `commandText` `kind`

ncdiehl11 commented on app/src/molecules/Command/CommandText.tsx:
```diff
@@ -169,3 +194,109 @@ function ThermocyclerRunProfile(
+                : profileElementTexts[0].cycleText}
+            </li>
+          ) : (
+            profileElementTexts.map((element, index: number) =>
```
👍

ncdiehl11 reviewed (commented):
Assessing this from a PD perspective, this looks good and shouldn't cause a big headache on our end. The main implication will be in `thermocyclerFormToArgs` to maintain a cycle profile step's dimensionality (and repetitions property) rather than flattening it into its comprising steps.

ecormany commented on app/src/assets/localization/en/protocol_command_text.json:
```diff
@@ -72,7 +72,9 @@
   "tc_awaiting_for_duration": "Waiting for Thermocycler profile to complete",
   "tc_run_profile_steps": "temperature: {{celsius}}°C, seconds: {{seconds}}",
-  "tc_starting_profile": "Thermocycler starting {{repetitions}} repetitions of cycle composed of the following steps:",
+  "tc_starting_extended_profile_cycle": "A cycle of {{repetitions}} repetitions of the following steps:",
+  "tc_starting_extended_profile": "Thermocycler starting {{elementCount}} element profile composed of the following elements:",
+  "tc_starting_profile": "Thermocycler starting {{stepCount}} step profile composed of the following:",
```
It's good timing for wordsmithing here, because I'm taking a look at our other command text across PD and `opentrons_simulate` as well.

My suggestions here are based on these general factors:
- We prefer "-ing" forms of verbs (which are present here already).
- We generally describe steps as doing something _to_ a module (direct object) rather than as the module doing something (subject). The existing TC step is an outlier in that regard.

In addition, I think that the top level should look very similar regardless of whether it has the flat or embedded structure.

```suggestion
  "tc_starting_extended_profile_cycle": "{{repetitions}} repetitions of the following steps:",
  "tc_starting_extended_profile": "Running thermocycler profile with {{elementCount}} total steps and cycles:",
  "tc_starting_profile": "Running thermocycler profile with {{stepCount}} steps:",
```

Something about line 74 strikes me as very "here is some data" instead of plain English, too. Pros and cons of going with something like `"{{seconds}} seconds at {{celsius}}°C"` instead:
- It sounds more like a human.
- Except it counts everything in seconds, and these could be long spans of time. Pop quiz: how long is 2700 seconds? Hours, minutes, seconds would be cool but probably a refactor for another day.
- Variable length cycles will make the output less tabular, and potentially harder to scan.

Bare minimum let's capitalize the T in `Temperature`, either in the string or with a transformation.

sfoster1 replied:
I like the suggested changes and will apply.

I think for changing the individual step texts, definitely sentence capsing, but I agree about the variable length thing, and temperatures should be a lot less variable. 

For timestamps, we can pretty trivially transform them to nicer durations - do you prefer `mm:ss`, `[mm minutes ] ss seconds`,  something else?

ecormany replied:
Two unambiguous options:
- `0:01:02`
-  `0h 01m 02s`

Even though it raises the question of "does this need to be i18n'd?" I like including the literal text `h`, `m`, `s` because it reduces mental burden of going "wait, which units is that colon separating?"

sfoster1 replied:
yeah, I agree, and yeah, it would need to be i18n'd but that's just incremental work and c'est la vie. Will do!

sfoster1 replied:
So that will be `Temperature: {temperature}°C, hold time: {hh}h {mm}m {ss}s`? Something else?

sfoster1 replied:
Actually, since it looks like we use durations in other places that we'll have to fix, let's just specify it as a duration in the i18n strings

ecormany reviewed (commented):
🫡 🔡 ⚒️ wordsmith reporting for duty

sfoster1 committed:
Update app/src/assets/localization/en/protocol_command_text.json

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

sfoster1 committed:
i should fix this

sfoster1 committed:
capital T and that rhymes with P and that stands for pool

sfoster1 committed:
some very specific tests

sfoster1 committed:
refactor(app): formatDuration improvements

This handles non-normalized durations (i.e., all seconds) now. I think
it is janky.

sfoster1 committed:
refactor(app): profile durations are nicer now

ecormany commented:
> for extremely long times, hours will just be extended - so you'd have `25h`, not `1d 1h`

I think if anything, times will tend to be on the short side of an hour. Nit, but a single `0` instead of `00` would probably suffice for any sub-hour time. If you get ragged text because you have a step that exceeds 10h, so be it.

sfoster1 committed:
single digit hours

sfoster1 committed:
whoops

sfoster1 committed:
whoops x2

sfoster1 commented:
![Untitled](https://github.com/user-attachments/assets/b9b79c04-0462-4c2f-b3b1-1c59e2d31623)

How's this @ecormany

ecormany reviewed (approved):
Command text is good. 🚢
